moveit_ros/moveit_servo/src/servo_calcs.cpp

```diff
@@ -732,60 +770,51 @@ double ServoCalcs::velocityScalingFactorForSingularity(const Eigen::VectorXd& co
+  // For current state, we know the current position well. For vel and accel, use the previous output from Reflexxes
+  // This assumes the hardware tracks the requested vel/accel well.
+  for (size_t joint_idx = 0; joint_idx < num_joints_; ++joint_idx)
   {
-    // Some joints do not have bounds defined
-    const auto bound = joint->getVariableBounds(joint->getName());
-
-    if (bound.velocity_bounded_)
-    {
-      velocity(joint_delta_index) = delta_theta(joint_delta_index) / parameters_->publish_period;
-
-      bool clip_velocity = false;
-      double velocity_limit = 0.0;
-      if (velocity(joint_delta_index) < bound.min_velocity_)
-      {
-        clip_velocity = true;
-        velocity_limit = bound.min_velocity_;
-      }
-      else if (velocity(joint_delta_index) > bound.max_velocity_)
-      {
-        clip_velocity = true;
-        velocity_limit = bound.max_velocity_;
-      }
-
-      // Apply velocity bounds
-      if (clip_velocity)
-      {
-        const double scaling_factor =
-            fabs(velocity_limit * parameters_->publish_period) / fabs(delta_theta(joint_delta_index));
+    reflexxes_position_input_param_->CurrentVelocityVector->VecData[joint_idx] =
+        reflexxes_position_output_param_->NewVelocityVector->VecData[joint_idx];
+    reflexxes_position_input_param_->CurrentAccelerationVector->VecData[joint_idx] =
+        reflexxes_position_output_param_->NewAccelerationVector->VecData[joint_idx];
+  }
+  reflexxes_wrapper::setCurrentPositions(reflexxes_position_input_param_.get(), num_joints_,
+                                         original_joint_state_.position);
 
-        // Store the scaling factor if it's the smallest yet
-        if (scaling_factor < velocity_limit_scaling_factor)
-          velocity_limit_scaling_factor = scaling_factor;
-      }
-    }
-    ++joint_delta_index;
+  // Update target state in Reflexxes
+  std::vector<double> target_positions = original_joint_state_.position;
+  for (size_t joint_idx = 0; joint_idx < num_joints_; ++joint_idx)
+  {
+    target_positions[joint_idx] += delta_theta[joint_idx];
+  }
+  Eigen::ArrayXd eigen_target_velocities = delta_theta / parameters_->publish_period;
+  std::vector<double> target_velocities(eigen_target_velocities.data(),
+                                        eigen_target_velocities.data() + eigen_target_velocities.size());
+  reflexxes_wrapper::setTargetState(reflexxes_position_input_param_.get(), num_joints_, target_positions,
+                                    target_velocities);
+
+  // Call the Reflexxes algorithm
+  int result = reflexxes_->RMLPosition(*reflexxes_position_input_param_, reflexxes_position_output_param_.get(),
+                                       reflexxes_flags_);
+
+  // Calculate the new delta_theta from Reflexxes output
+  for (size_t joint_idx = 0; joint_idx < num_joints_; ++joint_idx)
+  {
+    delta_theta[joint_idx] = reflexxes_position_output_param_->NewPositionVector->VecData[joint_idx] -
```

Comment:
Is this correct? You are actually calculating `delta_theta` from the step *k* and using it to set reference in step *k+1*.

I am not certain if this is plausible? Are you trying to keep the robot at its velocity limit?

Actually, thinking about this.... This could cause the robot moving backward if the frequency is not sufficient. When we used reflexxes before for the trajectory smoothing in controllers then we always set as goal current position reference (target state) and if needed, updated it at every control update. This leverage the trajectory generation of reflexxes and always worked very well. Maybe it is worth trying, or do you see some issue with that approach?

P.S. Knowing the author of reflexxes personally, I could check with him what is the right way to use the library.

Reply:
I'm pretty sure that is correct. It was copied from an example.